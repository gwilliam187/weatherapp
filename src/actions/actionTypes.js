// City Actions
export const ADD_CITY = 'ADD_CITY';
export const REMOVE_CITY = 'REMOVE_CITY';
export const TOGGLE_CITY_IS_PUBLIC = 'TOGGLE_CITY_IS_PUBLIC';
export const INITIALISE_CITY = "INITIALISE_CITY";

// Selected City Actions
export const SELECT_CITY = 'SELECT_CITY';
export const UNSELECT_CITY = 'UNSELECT_CITY';