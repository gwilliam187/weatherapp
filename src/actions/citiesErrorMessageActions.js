export const updateCitiesErrorMessage = (city, status) => {
	if(status === 'CITY_NOT_FOUND') {
		return {
			type: 'CITY_NOT_FOUND'
		};
	} else if(status === 'INVALID_INPUT') {
		return {
			type: 'INVALID_INPUT'
		};
	}

	return {
		type: 'ALLES_GUT'
	};
};