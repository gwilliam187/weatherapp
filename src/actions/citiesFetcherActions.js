import { 
	GET_CITIES_REQUEST, 
	GET_CITIES_SUCCESS, 
	GET_CITIES_FAILURE 
} from '../actions/actionTypes';

export const getCitiesRequest = () => {
	return {
		type: GET_CITIES_REQUEST,
	};
};

export const getCitiesSuccess = () => {
	return {
		type: GET_CITIES_SUCCESS,
	};
};

export const getCitiesFailure = () => {
	return {
		type: GET_CITIES_FAILURE,
	};
};
