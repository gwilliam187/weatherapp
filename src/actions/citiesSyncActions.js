import {
	CITIES_START_SYNC, 
	CITIES_STOP_SYNC,
} from '../actions/actionTypes';

export const citiesStartSync = () => {
	return {
		type: CITIES_START_SYNC,
	};
};

export const citiesStopSync = () => {
	return {
		type: CITIES_STOP_SYNC,
	};
};