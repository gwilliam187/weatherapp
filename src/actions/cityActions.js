export const selectCity = (city) => {
	return {
		type: 'SELECT_CITY',
		payload: city
	};
};

export const removeCity = (city) => {
	return {
		type: 'REMOVE_CITY',
		payload: city
	};
};