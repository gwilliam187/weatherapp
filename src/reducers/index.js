import { combineReducers } from 'redux';

import regionsReducer from './regionsReducer';
import selectedRegionsReducer from './selectedRegionReducer';
import citiesReducer from './citiesReducer';
import selectedCitiesReducer from './selectedCitiesReducer';
import weathersReducer from './weathersReducer';
import citiesErrorMessageReducer from './citiesErrorMessageReducer';
import userListReducer from './userListReducer';
import selectedUserReducer from './selectedUserReducer';
import loginReducer from './loginReducer';
import treeReducer from './treeReducer';
import rxdb from './rxdbReducer';
import selectedObjectReducer from './selectedObjectReducer';
import citiesFetcherReducer from './citiesFetcherReducer';
import citiesSyncReducer from './citiesSyncReducer';

export default combineReducers({
	regions: regionsReducer,
	selectedRegion: selectedRegionsReducer,
	users: userListReducer,
	selectedUser: selectedUserReducer,
	cities: citiesReducer,
	loginState: loginReducer,
	selectedCities: selectedCitiesReducer,
	weathers: weathersReducer,
	citiesErrorMessage: citiesErrorMessageReducer,
	trees: treeReducer,
	rxdb: rxdb,
	selectedObject: selectedObjectReducer,
	citiesFetcher: citiesFetcherReducer,
	citiesSync: citiesSyncReducer,
});