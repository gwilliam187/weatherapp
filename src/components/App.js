import React from 'react';
import { ToastContainer, toast, Flip } from 'react-toastify';
import { connect } from 'react-redux';
import 'react-toastify/dist/ReactToastify.css';

import { loadCities, updateCityToUser, addUser } from '../actions';
import { fetchWeathersForSelectedCities } from '../actions/weatherActions';
import UserList from './UserList';
import UserInput from './UserInput';
import UserBar from './UserBar';
import CityList from './CityList';
import CityInput from './CityInput';
import WeatherDisplay from './WeatherDisplay';
import DiffDisplay from './DiffDisplay';
import './custom-style.css';

class App extends React.Component {
	componentDidMount(){
		//this.props.addUser("steven_klarens");
		this.props.loadCities();
		// this.props.updateCityToUser();
	}

	componentDidUpdate() {
		this.props.fetchWeathersForSelectedCities();
	}

	renderSidebar() {
		if(this.props.selectedUser) {
			return (
				<div className='col-lg-4'>
					<div className='row'>
						<UserBar />
						<CityList />
						<CityInput />
					</div>
				</div>
			);
		} else {
			return (
				<div className='col-12 d-flex'>
					<div className='row flex-grow-1 align-self-end'>
						<UserInput />
					</div>
				</div>
			);
		}
	}

	renderWeatherDisplay() {
		if(this.state.cityList.length === 1) {
			return (
				<div className='row'>
					<div className='col-12 pb-3'>
						<WeatherDisplay />
					</div>
				</div>
			);
		} else if(this.state.cityList.length === 2) {
			return (
				<div className='row'>
					<div className='col-md-12 col-lg-6 pb-3'>
						<WeatherDisplay />
					</div>
					<div className='col-md-12 col-lg-6 pb-3'>
						<WeatherDisplay />
					</div>
				</div>
			);
		}
	}

	renderDiffDisplay() {
		if(this.state.tempList.length === 2) {
			return (
				<div className='row'>
					<div className='col-12'>
						<DiffDisplay tempList={this.state.tempList} windList={this.state.windList} />
					</div>
				</div>
			);
		}
	}

	render() {
		return (
			<div className="container app">
				<div className='row flex-grow-1'>
					{ this.renderSidebar() }
					<div className='col-lg-8'>
						<WeatherDisplay />
						<DiffDisplay />
					</div>
				</div>
				<ToastContainer
					toastClassName='custom-toast'
					position="top-right"
					autoClose={3000}
					hideProgressBar
					newestOnTop={false}
					closeOnClick
					rtl={false}
					pauseOnVisibilityChange
					draggable
					pauseOnHover
					transition={Flip}
				/>
			</div>
		);
	}
}

const mapStateToProps = state => {
	return {
		selectedCities: state.selectedCities,
		selectedUser: state.selectedUser
	};
};

const mapDispatchToProps = {
	fetchWeathersForSelectedCities, 
	loadCities, 
	updateCityToUser,
	addUser
};

export default connect(mapStateToProps, mapDispatchToProps)(App);