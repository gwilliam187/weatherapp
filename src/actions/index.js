import _ from 'lodash';

import openWeather from '../apis/openWeatherAPI';
import { selectCity, removeCity } from './cityActions';
import { fetchWeather } from './weatherActions';

// City Actions
export { selectCity };
export { removeCity };

// Weather Actions
export { fetchWeather };


export const fetchWeathersForSelectedCities = () => async (dispatch, getState) => {
	const selectedCities = getState().selectedCities;
	const promises = await selectedCities.map(async selectedCity => {
		const res = await dispatch(fetchWeather(selectedCity));
		return res;
	});

	Promise.all(promises)
		.then(res => {
			let error = {};
			const weathers = res
				.map(currRes => {
					if('status' in currRes) {
						return currRes.data
					} else {
						dispatch(removeCity(currRes.city));
						error.city = currRes.city;
						return undefined;
					}
				})
				.filter(weather => {
					return weather !== undefined;
				});

			dispatch({
				type: 'FETCH_WEATHERS',
				payload: weathers
			});
			//setWeathers(weathers)(dispatch, getState);

			if(!_.isEmpty(error)) {
				dispatch(updateCitiesErrorMessage(error.city, 'CITY_NOT_FOUND'));
			}
		})
		.catch(err => { 
			console.log(err); 
		});
};

export const updateCitiesErrorMessage = (city, status) => {
	if(status === 'CITY_NOT_FOUND') {
		console.log('returning not found');
		return {
			type: 'CITY_NOT_FOUND',
			payload: city.cityName
		};
	} else if(status === 'INVALID_INPUT') {
		return {
			type: 'INVALID_INPUT'
		};
	}

	return {
		type: 'ALLES_GUT'
	};
};