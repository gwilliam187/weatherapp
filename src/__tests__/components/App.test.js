import React from 'react';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import _ from 'lodash';
// import axios from 'axios';
// jest.mock('axios');

import App from '../../components/App';
import reducers from '../../reducers';
import successOpenWeatherJson from '../json/successOpenWeatherAPI';
import { fetchWeathersForSelectedCities } from '../../actions';

function setupStore() {
  return createStore(reducers, applyMiddleware(thunk));
}

const flushAllPromises = () => {
	return new Promise(resolve => setImmediate(resolve));
}

const mockAxios = {
  get: jest.fn().mockImplementation(() => {
    return new Promise((res, rej) => {
      mockAxios._resolve = res;
      mockAxios._reject = rej;
    });
  }),
  mockSuccess (resp) {
    this._resolve(resp);
  },
  mockError (resp) {
    this._reject(resp);
  },
  _resolve: null,
  _reject: null
};

describe('<App /> with Redux', () => {
	let wrapper;
	let store;

	beforeEach(() => {
		store = setupStore();
		wrapper = mount(
			<Provider store={store}><App /></Provider>
    );
	});

	test('Initializes successfully', () => {
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	test('Selecting Cities', async () => {
		const cityButton1 = wrapper.find('CityButton').at(0);
		const button1 = cityButton1.find('button');
		button1.simulate('click');

		const cityButton2 = wrapper.find('CityButton').at(1);
		const button2 = cityButton2.find('button');
		button2.simulate('click');
		const expectedState = 
		[
			{
				cityName: cityButton1.props().cityName,
				countryCode: cityButton1.props().countryCode		
			},
			{
				cityName: cityButton2.props().cityName,
				countryCode: cityButton2.props().countryCode		
			}
		]
		expect(store.getState().selectedCities).toEqual(expectedState);

		const selectedCities = store.getState().selectedCities;
		const promises = await selectedCities.map(async selectedCity => {
			const promise = mockAxios.get();
			mockAxios.mockSuccess({ status: 200, data: successOpenWeatherJson });
			return promise;
		});
		Promise.all(promises)
			.then(res => {
				let error = {};
				const weathers = res
					.map(currRes => {
						return currRes.data;
					});

				store.dispatch({
					type: 'FETCH_WEATHERS',
					payload: weathers
				});

				// console.log(weathers);
				// console.log(store.getState().weathers);
				expect(store.getState().weathers).toEqual(weathers);

				if(!_.isEmpty(error)) {
					store.dispatch(updateCitiesErrorMessage(error.city, 'CITY_NOT_FOUND'));
				}
				expect(store.getState().citiesErrorMessage).toEqual('');
				
			})
			.catch(err => {console.log(err)});
	});
});