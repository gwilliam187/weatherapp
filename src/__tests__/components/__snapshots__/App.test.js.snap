// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> with Redux Initializes successfully 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(App)>
    <App
      fetchWeathersForSelectedCities={[Function]}
      selectedCities={Array []}
    >
      <div
        className="container app"
      >
        <div
          className="row w-100"
        >
          <div
            className="col-4 shadow city-button-wrapper"
          >
            <div
              className="row header"
            >
              <div
                className="col-12 my-4"
              >
                <h1>
                  Cities
                </h1>
              </div>
              <Connect(CityList)>
                <CityList
                  cities={
                    Array [
                      Object {
                        "cityName": "Berlin",
                        "countryCode": "de",
                      },
                      Object {
                        "cityName": "Potsdam",
                        "countryCode": "de",
                      },
                      Object {
                        "cityName": "Soest",
                        "countryCode": "de",
                      },
                      Object {
                        "cityName": "Jakarta",
                        "countryCode": "id",
                      },
                      Object {
                        "cityName": "Bangkok",
                        "countryCode": "th",
                      },
                      Object {
                        "cityName": "Moscow",
                        "countryCode": "ru",
                      },
                    ]
                  }
                  dispatch={[Function]}
                >
                  <div
                    className="col-12"
                  >
                    <Connect(CityButton)
                      cityName="Berlin"
                      countryCode="de"
                      key="Berlin"
                    >
                      <CityButton
                        cityName="Berlin"
                        countryCode="de"
                        selectCity={[Function]}
                        updateCitiesErrorMessage={[Function]}
                      >
                        <div
                          className="py-2"
                        >
                          <button
                            className="btn btn-outline-secondary btn-lg city-button"
                            onClick={[Function]}
                          >
                            Berlin
                          </button>
                        </div>
                      </CityButton>
                    </Connect(CityButton)>
                    <Connect(CityButton)
                      cityName="Potsdam"
                      countryCode="de"
                      key="Potsdam"
                    >
                      <CityButton
                        cityName="Potsdam"
                        countryCode="de"
                        selectCity={[Function]}
                        updateCitiesErrorMessage={[Function]}
                      >
                        <div
                          className="py-2"
                        >
                          <button
                            className="btn btn-outline-secondary btn-lg city-button"
                            onClick={[Function]}
                          >
                            Potsdam
                          </button>
                        </div>
                      </CityButton>
                    </Connect(CityButton)>
                    <Connect(CityButton)
                      cityName="Soest"
                      countryCode="de"
                      key="Soest"
                    >
                      <CityButton
                        cityName="Soest"
                        countryCode="de"
                        selectCity={[Function]}
                        updateCitiesErrorMessage={[Function]}
                      >
                        <div
                          className="py-2"
                        >
                          <button
                            className="btn btn-outline-secondary btn-lg city-button"
                            onClick={[Function]}
                          >
                            Soest
                          </button>
                        </div>
                      </CityButton>
                    </Connect(CityButton)>
                    <Connect(CityButton)
                      cityName="Jakarta"
                      countryCode="id"
                      key="Jakarta"
                    >
                      <CityButton
                        cityName="Jakarta"
                        countryCode="id"
                        selectCity={[Function]}
                        updateCitiesErrorMessage={[Function]}
                      >
                        <div
                          className="py-2"
                        >
                          <button
                            className="btn btn-outline-secondary btn-lg city-button"
                            onClick={[Function]}
                          >
                            Jakarta
                          </button>
                        </div>
                      </CityButton>
                    </Connect(CityButton)>
                    <Connect(CityButton)
                      cityName="Bangkok"
                      countryCode="th"
                      key="Bangkok"
                    >
                      <CityButton
                        cityName="Bangkok"
                        countryCode="th"
                        selectCity={[Function]}
                        updateCitiesErrorMessage={[Function]}
                      >
                        <div
                          className="py-2"
                        >
                          <button
                            className="btn btn-outline-secondary btn-lg city-button"
                            onClick={[Function]}
                          >
                            Bangkok
                          </button>
                        </div>
                      </CityButton>
                    </Connect(CityButton)>
                    <Connect(CityButton)
                      cityName="Moscow"
                      countryCode="ru"
                      key="Moscow"
                    >
                      <CityButton
                        cityName="Moscow"
                        countryCode="ru"
                        selectCity={[Function]}
                        updateCitiesErrorMessage={[Function]}
                      >
                        <div
                          className="py-2"
                        >
                          <button
                            className="btn btn-outline-secondary btn-lg city-button"
                            onClick={[Function]}
                          >
                            Moscow
                          </button>
                        </div>
                      </CityButton>
                    </Connect(CityButton)>
                  </div>
                </CityList>
              </Connect(CityList)>
              <Connect(CityInput)>
                <CityInput
                  citiesErrorMessage=""
                  selectCity={[Function]}
                  updateCitiesErrorMessage={[Function]}
                >
                  <div
                    className="col-12 py-2"
                  >
                    <input
                      className="form-control form-control-lg"
                      onChange={[Function]}
                      onKeyDown={[Function]}
                      placeholder="e.g. Berlin,de"
                      type="text"
                      value=""
                    />
                    <div
                      className="text-left text-danger mt-1"
                    />
                  </div>
                </CityInput>
              </Connect(CityInput)>
            </div>
          </div>
          <div
            className="col-8"
          >
            <Connect(WeatherDisplay)>
              <WeatherDisplay
                fetchWeathersForSelectedCities={[Function]}
                weathers={Array []}
              >
                <div />
              </WeatherDisplay>
            </Connect(WeatherDisplay)>
            <Connect(DiffDisplay)>
              <DiffDisplay
                dispatch={[Function]}
                weathers={Array []}
              >
                <div />
              </DiffDisplay>
            </Connect(DiffDisplay)>
          </div>
        </div>
      </div>
    </App>
  </Connect(App)>
</Provider>
`;
